pkg/http/listener.go

```diff
@@ -0,0 +1,326 @@
+	for _, method := range methods {
+		if s == method {
+			return true
+		}
+	}
+
+	return false
+}
+
+type acceptResult struct {
+	conn net.Conn
+	err  error
+}
+
+// httpListener - HTTP listener capable of handling multiple server addresses.
+type httpListener struct {
+	mutex                  sync.Mutex         // to guard Close() method.
+	tcpListeners           []*net.TCPListener // underlaying TCP listeners.
+	acceptCh               chan acceptResult  // channel where all TCP listeners write accepted connection.
+	doneChs                []chan struct{}    // done channels for each goroutine for TCP listeners.
+	tlsConfig              *tls.Config        // TLS configuration
+	tcpKeepAliveTimeout    time.Duration
+	readTimeout            time.Duration
+	writeTimeout           time.Duration
+	updateBytesReadFunc    func(int)                           // function to be called to update bytes read in bufConn.
+	updateBytesWrittenFunc func(int)                           // function to be called to update bytes written in bufConn.
+	errorLogFunc           func(error, string, ...interface{}) // function to be called on errors.
+}
+
+// start - starts separate goroutine for each TCP listener.  A valid insecure/TLS HTTP new connection is passed to httpListener.acceptCh.
+func (listener *httpListener) start() {
+	listener.acceptCh = make(chan acceptResult)
+
+	// Closure to send acceptResult to acceptCh.
+	// It returns true if the result is sent else false if returns when doneCh is closed.
+	send := func(result acceptResult, doneCh <-chan struct{}) bool {
+		select {
+		case listener.acceptCh <- result:
+			// Successfully written to acceptCh
+			return true
+		case <-doneCh:
+			// As stop signal is received, close accepted connection.
+			if result.conn != nil {
+				result.conn.Close()
+			}
+			return false
+		}
+	}
+
+	// Closure to accept single connection.
+	acceptTCP := func(tcpConn *net.TCPConn, doneCh <-chan struct{}) {
+		// Tune accepted TCP connection.
+		tcpConn.SetKeepAlive(true)
+		tcpConn.SetKeepAlivePeriod(listener.tcpKeepAliveTimeout)
+
+		bufconn := newBufConn(tcpConn, listener.readTimeout, listener.writeTimeout,
+			listener.updateBytesReadFunc, listener.updateBytesWrittenFunc)
+
+		// Peek bytes of maximum length of all HTTP methods.
+		data, err := bufconn.Peek(methodMaxLen)
+		if err != nil {
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(err,
+					"Error in reading from new connection %s at server %s",
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			bufconn.Close()
+			return
+		}
+
+		// Return bufconn if read data is a valid HTTP method.
+		tokens := strings.SplitN(string(data), " ", 2)
+		if isHTTPMethod(tokens[0]) {
+			if listener.tlsConfig == nil {
+				send(acceptResult{bufconn, nil}, doneCh)
+			} else {
+				// As TLS is configured and we got plain text HTTP request,
+				// return 403 (forbidden) error.
+				bufconn.Write(sslRequiredErrMsg)
+				bufconn.Close()
+			}
+			return
+		}
+
```

Comment:
I am not able to understand.  Below is the logic of this block.

1. Tune `TCPConn` and make `bufConn`
2. Peek bytes of maximum length of all HTTP methods irrespective of TLS configuration.
3. If valid HTTP method is found and TLS is not configured, we got valid client connection and return.
4. If valid HTTP method is found and TLS is configured, return error `sslRequiredErrMsg` indicating client should use TLS and drop the connection.
5. If invalid HTTP method is found and TLS is configured, attempt for TLS handshake.     
  i. If handshake fails, drop the connection. may be junk data sent by the client. 
  ii. If handshake succeeds, peek bytes of maximum length of all HTTP methods to check whether the client sends valid HTTP request via TLS. If its valid, we got valid client connection and return
6.  Finally drop the connection because of junk data is sent by the client in TLS/non-TLS mode.

With this logic, I don't find the leak problem in the implementation.